### PR TITLE
Freshwater crisis fix

### PR DESCRIPTION
**Pull request by yairm210**
Fresh water is currently a mess, as outlined in #6186 
Step 1 of resolving #6186 freshwater crisis is to unify adjacency to a single function, no more isAdjacentToFreshwater.

Remaining problems, not part of PR but here for discussion:
 - "[+1 Food] from [Fresh Water] tiles" and even worse, "[+1 Food] from [non-freshwater] tiles" - replaced with new tile conditionals, \<in tiles adjacent to [Fresh water]> and \<in tiles not adjacent to [Fresh water]>
 - "Can also be built on tiles adjacent to fresh water" that enables farms on *hills* can be replaced with *enabling* hills and adding "Cannot be built on [tileFilter] tiles <in tiles not adjacent to [Fresh water]>"
 
That should get rid of existing problems while also adding adjacency conditionals and removing an extremely badly-worded unique

### CODE REVIEW

**Comment by SomeTroglodyte** [comment]
So `ConditionalInTilesNot` was defined but not actually implemented before? Strange.
And - will this count an Oasis itself as having fresh H<sub>2</sub>O? It should, right?

**Comment by yairm210** [comment]
Yes, and Oasis has the "Fresh water" unique, so yes.

File: core/src/com/unciv/logic/map/TileInfo.kt [hunk 1]
```diff
@@ -528,14 +522,14 @@ open class TileInfo {
...
             improvement.terrainsCanBeBuiltOn.isEmpty() && improvement.turnsToBuild == 0 && isLand -> true
             improvement.terrainsCanBeBuiltOn.contains(topTerrain.name) -> true
-            improvement.uniqueObjects.filter { it.type == UniqueType.MustBeNextTo }.any {
+            improvement.getMatchingUniques(UniqueType.MustBeNextTo).any {
                 !isAdjacentTo(it.params[0])
             } -> false
             !isWater && RoadStatus.values().any { it.name == improvement.name && it > roadStatus } -> true
             improvement.name == roadStatus.removeAction -> true
             topTerrain.unbuildable && !improvement.isAllowedOnFeature(topTerrain.name) -> false
             // DO NOT reverse this &&. isAdjacentToFreshwater() is a lazy which calls a function, and reversing it breaks the tests.
```
Comment (SomeTroglodyte) [inline core/src/com/unciv/logic/map/TileInfo.kt:531]: Comment obsolete

**Review by SomeTroglodyte — APPROVED** [review]
I didn't test, but can't see how this could go wrong

**Comment by xlenstra** [comment]
> "Can also be built on tiles adjacent to fresh water" that enables farms on hills can be replaced with enabling hills and adding "Cannot be built on [tileFilter] tiles <in tiles not adjacent to [Fresh water]>"

Doesn't the current setup also allow building farms on snow, tundra, desert (next to oasis), etc., which would then be lost? I'm not sure what the original does, but at least for modding deprecating that unique would be a loss.
Additionally, "[...] on hills tiles in tiles adjacent [...]" to me sounds a bit weird, with  'tiles' being in the same sentence so close together. Maybe make a new unique "adjacent to [tileFilter]" that behaves the same so modders can choose which sounds better? Or just fix this in `uniqueReplacementText` or however it is called?

File: core/src/com/unciv/logic/map/TileInfo.kt [hunk 2]
```diff
@@ -574,7 +568,7 @@ open class TileInfo {
             "Water resource" -> isWater && observingCiv != null && hasViewableResource(observingCiv)
             "Natural Wonder" -> naturalWonder != null
             "Featureless" -> terrainFeatures.isEmpty()
-            "Fresh Water" -> isAdjacentToFreshwater
+            "Fresh Water" -> isAdjacentTo(Constants.freshWater)
```
Comment (xlenstra) [inline core/src/com/unciv/logic/map/TileInfo.kt:571]: So we still have both "Fresh water" and "Fresh Water" and support only one or both depending on the unique?

Comment (SomeTroglodyte) [reply core/src/com/unciv/logic/map/TileInfo.kt:571]: How about `const`ing that one as well, then naming one freshWaterSource and the other freshWaterAccess (the latter the one with capital Water), irrespective of actual string value - so the uses are traceable and changing the actual keys, if at all, becomes more readable?

Comment (yairm210) [reply core/src/com/unciv/logic/map/TileInfo.kt:571]: The goal is to remove all support for "Fresh Water", but this will I think require adding a warning in the ruleset checks

**Review by xlenstra — APPROVED** [review]

**Comment by SomeTroglodyte** [comment]
Some more loud thinking:
- Capitalized "Fresh Water" is now _only_  used in the conditional of the resource generation uniques, only implemented in `TileInfo.matchesTerrainFilter` and allowed by `UniqueParameterType.TerrainFilter.knownValues` - consistent and clean. A comment pointing out the special role of the "W" in matchesTerrainFilter might help.
- Gardens (and Floating Gardens) - will use `cityCenter.isAdjacentTo("Fresh water")` which, in absence of rivers, will allow the building next to a lake or oasis but not if the city is _on_ the source and no neighbor has one. Petra issue again.
- Farm (and Terrace farm) - uses `ImprovementStatsOnTile` - the one with the "endless feedback" comment. The question of the center tile is moot, no Farm on Oasis...
- Lower case "Fresh water" now means "this tile is a source" or "check for access to" depending on circumstances. A Tilefilter testing whether "this" tile is a source without looking at the neighbors is no longer possible.
- The `FreshWater`UniqueType could use the const, too

Tldr - more analysis and my approval still stands.

I'd say we need unit tests for a few typical vanilla ruleset situations - There's already one test afaik that loads an entire map from a gzipped string, we could go that route...

**Comment by SomeTroglodyte** [comment]
<details><summary>Toying with this <code>FreshWaterTests.kt</code>:</summary>

```kotlin
package com.unciv.logic.map

import com.unciv.Constants
import com.unciv.logic.MapSaver
import com.unciv.models.ruleset.Ruleset
import com.unciv.models.ruleset.RulesetCache
import com.unciv.testing.GdxTestRunner
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(GdxTestRunner::class)
class FreshWaterTests {
    private var map = TileMap()
    private var ruleSet = Ruleset()

    companion object {
        private const val testingMap = "H4sIAAAAAAAA/51TvU7DMBB+lTyAOzSI5UYKFUOkooDKUHUw5CAnnDiyz4Vg5d1xlIn8tc0SyfH3p/vOvpDVkzSyQEZjwXNdITwUFdci3DzTL4I3MiNnIRbflHEOtyJH+swZbkQZeLBxlnXRiDdpMXUKLTJs6BTto1W0lyUpJRvBpDAhy3DwLe4FjZFUwn1AGxYGrXbmHeE1R8mN8JW2xKRL8D+wWos6fDqD/8QesgWex02JzaWY0J0PvoByfbLAidv5xKNE7o7bwHJBBA47ackeewqtQI+fyC+0o0VcaPQYeu/7rIc+l9MnrWdru5rSpZzoIZf2TnPY9gQ/OKUTGmDjcGxOZwTS9gVNKixIff1wut7Hcw762Cqts6hS4edgfboFXCR0bP4A33i02IEEAAA="
    }

    @Before
    fun initTheWorld() {
        RulesetCache.loadRulesets()
        ruleSet = RulesetCache.getVanillaRuleset()
        map = MapSaver.mapFromSavedString(testingMap, false)
        map.setTransients(ruleSet, false)
    }

    @Test
    fun isAdjacentToFreshWater() {
        for (tile in map.values) {
            val isFresh = tile.isAdjacentTo(Constants.freshWater)
            val hasWheat = tile.resource == "Wheat"
            Assert.assertFalse("Tile $tile has fresh water but should not", isFresh && hasWheat)
            Assert.assertFalse("Tile $tile should have fresh water but doesn't", !isFresh && !hasWheat)
        }
    }

    fun canBuildFarm() {
        //TODO
    }

    fun canBuildGarden() {
        //TODO
    }
}
```
</details>
<details><summary>Which uses this map:</summary>

image
</details>
(I misused Wheat as a marker for tiles that shouldn't have fresh water)

...Fails because loading a map and setTransients on it will still leave all the `terrainFeatureObjects` empty, and if one fixes that, then...
